File: app/src/main/java/com/karol/sezonnazdrowie/model/SnzAlarmManager.java
package com.karol.sezonnazdrowie.model;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.preference.PreferenceManager;
import android.util.Log;

import com.karol.sezonnazdrowie.R;
import com.karol.sezonnazdrowie.data.Database;
import com.karol.sezonnazdrowie.data.FoodItem;
import com.karol.sezonnazdrowie.view.controls.TimePreference;
import com.prolificinteractive.materialcalendarview.CalendarDay;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

public class SnzAlarmManager {

    private final static String TAG = "SNZALARMMANAGER";

    public static void startSetAlarmsTask(final Context ctx, final Database database) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                SnzAlarmManager.setAlarms(ctx, database);
                return null;
            }
        }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    static void setAlarms(Context ctx, Database database) {
        HashMap<CalendarDay, ArrayList<FoodItem>> startMap = new HashMap<>();
        HashMap<CalendarDay, ArrayList<FoodItem>> endMap = new HashMap<>();

        fillMapWithItems(startMap, endMap, database.getAllFruits());
        fillMapWithItems(startMap, endMap, database.getAllVegetables());

        ArrayList<Integer> startDays = new ArrayList<>();
        Set<String> seasonStart = PreferenceManager.getDefaultSharedPreferences(ctx).getStringSet("pref_season_start", null);
        if (seasonStart == null)
            startDays.add(7);
        else {
            if (seasonStart.contains(ctx.getString(R.string.at_the_start_day)))
                startDays.add(0);
            if (seasonStart.contains(ctx.getString(R.string.week_before)))
                startDays.add(7);
            if (seasonStart.contains(ctx.getString(R.string.month_before)))
                startDays.add(30);
        }

        Intent intent;
        PendingIntent alarmIntent;
        AlarmManager alarmManager = (AlarmManager) ctx.getSystemService(Context.ALARM_SERVICE);
        int reqCode = 1;
        String notiTime = PreferenceManager.getDefaultSharedPreferences(ctx).getString("pref_notification_hour", "20:00");
        int notiHour = TimePreference.getHour(notiTime);
        int notiMinute = TimePreference.getMinute(notiTime);

        Calendar today = Calendar.getInstance();
        for (CalendarDay day : startMap.keySet()) {
			StringBuilder strBuilder = new StringBuilder();
			for (FoodItem item : startMap.get(day)) {
				if (PreferenceManager.getDefaultSharedPreferences(ctx).getBoolean("pref_noti_" + item.getName(), true)) {
					if (strBuilder.length() > 0)
						strBuilder.append(", ");
					strBuilder.append(item.getConjugatedName());
				}
			}
			if (strBuilder.length() > 0) {
	            for (Integer dayDiff : startDays) {
					String title = "";
					if (dayDiff == 0)
						title = ctx.getString(R.string.season_starts_soon);
					else if (dayDiff == 7)
						title = ctx.getString(R.string.season_starts_week);
					else if (dayDiff == 30)
						title = ctx.getString(R.string.season_starts_month);
					intent = new Intent(ctx, Receiver.class);
					intent.putExtra("type", "start");
                    intent.putExtra("reqCode", reqCode);
					intent.putExtra("title", title);
					intent.putExtra("text", strBuilder.toString());
					alarmIntent = PendingIntent.getBroadcast(ctx, reqCode++, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    	            Calendar calendar = (Calendar) day.getCalendar().clone();
        	        calendar.set(Calendar.YEAR, today.get(Calendar.YEAR));
            	    calendar.add(Calendar.DAY_OF_MONTH, -dayDiff);
	                if (calendar.before(today))
	 	                calendar.add(Calendar.YEAR, 1);
	                calendar.set(Calendar.HOUR_OF_DAY, notiHour);
	                calendar.set(Calendar.MINUTE, notiMinute);
	                calendar.set(Calendar.SECOND, 0);

	                alarmManager.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), alarmIntent);
                    Log.d(TAG, "setAlarms: Start Alarm set @ " + calendar.get(Calendar.DAY_OF_MONTH) + "." + calendar.get(Calendar.MONTH));
                }
			}
        }

        ArrayList<Integer> endDays = new ArrayList<>();
        Set<String> seasonEnd = PreferenceManager.getDefaultSharedPreferences(ctx).getStringSet("pref_season_end", null);
        if (seasonEnd == null)
            endDays.add(7);
        else {
            if (seasonEnd.contains(ctx.getString(R.string.at_the_end_day)))
                endDays.add(0);
            if (seasonEnd.contains(ctx.getString(R.string.week_before)))
                endDays.add(7);
        }

        for (CalendarDay day : endMap.keySet()) {
			StringBuilder strBuilder = new StringBuilder();
			for (FoodItem item : endMap.get(day)) {
                if (PreferenceManager.getDefaultSharedPreferences(ctx).getBoolean("pref_noti_" + item.getName(), true)) {
					if (strBuilder.length() > 0)
						strBuilder.append(", ");
					strBuilder.append(item.getConjugatedName());
				} else
                    Log.d(TAG, "setAlarms: Skipping the item " + item.getName());
			}
			if (strBuilder.length() > 0) {
	            for (Integer dayDiff : endDays) {
					String title = "";
					if (dayDiff == 0)
						title = ctx.getString(R.string.season_ends_soon);
					else if (dayDiff == 7)
						title = ctx.getString(R.string.season_ends_week);
					intent = new Intent(ctx, Receiver.class);
					intent.putExtra("type", "end");
                    intent.putExtra("reqCode", reqCode);
					intent.putExtra("title", title);
					intent.putExtra("text", strBuilder.toString());
					alarmIntent = PendingIntent.getBroadcast(ctx, reqCode++, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    	            Calendar calendar = (Calendar) day.getCalendar().clone();
        	        calendar.set(Calendar.YEAR, today.get(Calendar.YEAR));
        	        calendar.add(Calendar.DAY_OF_MONTH, -dayDiff);
        	        if (calendar.before(today))
        	            calendar.add(Calendar.YEAR, 1);
        	        calendar.set(Calendar.HOUR_OF_DAY, notiHour);
        	        calendar.set(Calendar.MINUTE, notiMinute);
        	        calendar.set(Calendar.SECOND, 0);
	
	                alarmManager.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), alarmIntent);
                    Log.d(TAG, "setAlarms: End Alarm set @ " + calendar.get(Calendar.DAY_OF_MONTH) + "." + calendar.get(Calendar.MONTH));
				}
            }
        }
        int prevMaxReqCode = PreferenceManager.getDefaultSharedPreferences(ctx).getInt("maxReqCode", 0);
        Log.d(TAG, "setAlarms: Ending with " + reqCode + ", previous ending " + prevMaxReqCode);
        PreferenceManager.getDefaultSharedPreferences(ctx).edit().putInt("maxReqCode", reqCode).apply();

        intent = new Intent(ctx, Receiver.class);
        while (prevMaxReqCode > reqCode) {
            Log.d(TAG, "setAlarms: Cancelling an alarm of reqCode = " + reqCode);
            alarmIntent = PendingIntent.getBroadcast(ctx, reqCode++, intent, PendingIntent.FLAG_UPDATE_CURRENT);
            alarmManager.cancel(alarmIntent);
        }
        PreferenceManager.getDefaultSharedPreferences(ctx).edit().putBoolean("pref_alarms_set", true).apply();
    }

    private static void fillMapWithItems(
            HashMap<CalendarDay, ArrayList<FoodItem>> startMap,
            HashMap<CalendarDay, ArrayList<FoodItem>> endMap,
            List<FoodItem> items) {
        for (FoodItem item : items) {
            if (item.getStartDay1() == null)
                continue;

            if (startMap.containsKey(item.getStartDay1()))
                startMap.get(item.getStartDay1()).add(item);
            else {
                ArrayList<FoodItem> list = new ArrayList<>();
                list.add(item);
                startMap.put(item.getStartDay1(), list);
            }
            if (item.getStartDay2() != null) {
                if (startMap.containsKey(item.getStartDay2()))
                    startMap.get(item.getStartDay2()).add(item);
                else {
                    ArrayList<FoodItem> list = new ArrayList<>();
                    list.add(item);
                    startMap.put(item.getStartDay2(), list);
                }
            }

            if (endMap.containsKey(item.getEndDay1()))
                endMap.get(item.getEndDay1()).add(item);
            else {
                ArrayList<FoodItem> list = new ArrayList<>();
                list.add(item);
                endMap.put(item.getEndDay1(), list);
            }
            if (item.getEndDay2() != null) {
                if (endMap.containsKey(item.getEndDay2()))
                    endMap.get(item.getEndDay2()).add(item);
                else {
                    ArrayList<FoodItem> list = new ArrayList<>();
                    list.add(item);
                    endMap.put(item.getEndDay2(), list);
                }
            }
        }
    }
}

File: app/src/main/java/com/karol/sezonnazdrowie/model/SnzAlarmManager.kt
package com.karol.sezonnazdrowie.model

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.annotation.StringRes
import com.karol.sezonnazdrowie.R
import com.karol.sezonnazdrowie.data.Database
import com.karol.sezonnazdrowie.data.FoodItem
import com.karol.sezonnazdrowie.data.SnzDatabase
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalDateTime
import org.threeten.bp.LocalTime
import org.threeten.bp.MonthDay
import org.threeten.bp.OffsetDateTime
import org.threeten.bp.ZoneOffset
import java.util.concurrent.Executors

object SnzAlarmManager {
    private const val TAG = "SNZALARMMANAGER"

    fun startSetAlarmsTask(ctx: Context, database: SnzDatabase) {
        val currentDayProvider = TimeDataProviderImpl()
        val persistentStorage: PersistentStorage = SharedPreferencesPersistentStorage(ctx)
        val alarmManager = ctx.getSystemService(Context.ALARM_SERVICE) as AlarmManager

        Executors.newSingleThreadExecutor().submit {
            setAlarms(ctx, currentDayProvider, persistentStorage, alarmManager, database)
        }
    }

    enum class PreNotiTimePeriod(val prefValue: String) {
        AtDay("DAY"),
        WeekBefore("WEEK"),
        MonthBefore("MONTH");

        companion object {
            fun fromPrefValue(requestedValue: String) =
                values().firstOrNull { it.prefValue == requestedValue }
                    ?: throw IllegalArgumentException("Unknown period of '$requestedValue'")
        }
    }

    interface TimeDataProvider {
        fun getCurrentDay(): LocalDate
        fun getZoneOffset(): ZoneOffset
    }

    class TimeDataProviderImpl : TimeDataProvider {
        override fun getCurrentDay(): LocalDate = LocalDate.now()
        override fun getZoneOffset(): ZoneOffset = OffsetDateTime.now().offset
    }

    private interface PrecedingPeriodObtainer {
        fun getPrecedingPeriod(): List<PreNotiTimePeriod>
    }

    private class StartPrecedingDaysObtainer(
        private val persistentStorage: PersistentStorage
    ) : PrecedingPeriodObtainer {

        override fun getPrecedingPeriod(): List<PreNotiTimePeriod> =
            persistentStorage.seasonStartPeriods.map { PreNotiTimePeriod.fromPrefValue(it) }
    }

    private class EndPrecedingDaysObtainer(
        private val persistentStorage: PersistentStorage
    ) : PrecedingPeriodObtainer {

        override fun getPrecedingPeriod(): List<PreNotiTimePeriod> =
            persistentStorage.seasonEndPeriods.map { PreNotiTimePeriod.fromPrefValue(it) }
    }

    private interface PrecedingPeriodMapper {
        fun mapToStringResource(daysAmount: PreNotiTimePeriod): Int
    }

    private class StartPrecedingDaysMapper : PrecedingPeriodMapper {

        override fun mapToStringResource(daysAmount: PreNotiTimePeriod): Int {
            return when (daysAmount) {
                PreNotiTimePeriod.AtDay -> R.string.season_starts_soon
                PreNotiTimePeriod.WeekBefore -> R.string.season_starts_week
                PreNotiTimePeriod.MonthBefore -> R.string.season_starts_month
            }
        }
    }

    private class EndPrecedingDaysMapper : PrecedingPeriodMapper {

        override fun mapToStringResource(daysAmount: PreNotiTimePeriod): Int {
            return when (daysAmount) {
                PreNotiTimePeriod.AtDay -> R.string.season_ends_soon
                PreNotiTimePeriod.WeekBefore -> R.string.season_ends_week
                PreNotiTimePeriod.MonthBefore -> throw IllegalArgumentException("Month is not supported for end dates")
            }
        }
    }

    /**
     * @return amount of alarms created
     */
    private fun createAlarmsFor(
        timeDataProvider: TimeDataProvider,
        map: Map<MonthDay, List<FoodItem>>,
        precedingPeriodObtainer: PrecedingPeriodObtainer,
        precedingDaysMapper: PrecedingPeriodMapper,
        isStart: Boolean,
        initialAlarmId: Int,
        ctx: Context,
        persistentStorage: PersistentStorage,
        alarmManager: AlarmManager
    ): Int {
        val today = timeDataProvider.getCurrentDay()
        val zoneOffset = timeDataProvider.getZoneOffset()

        var alarmId = initialAlarmId

        val notiTime = persistentStorage.notificationTime

        val precedingPeriodList = precedingPeriodObtainer.getPrecedingPeriod()

        map.entries.forEach {
            val notificationText = it.value
                .filter { foodItem ->
                    persistentStorage.isFoodItemNotificationEnabled(foodItem)
                }.joinToString { foodItem -> foodItem.conjugatedName }

            if (notificationText.isNotEmpty()) {
                daysIter@ for (precedingPeriod in precedingPeriodList) {
                    val titleRes: Int
                    try {
                        titleRes = precedingDaysMapper.mapToStringResource(precedingPeriod)
                    } catch (ex: IllegalArgumentException) {
                        continue@daysIter
                    }
                    val alarmIntent = createBroadcastNotificationIntent(
                        ctx,
                        isStart,
                        alarmId++,
                        titleRes,
                        notificationText
                    )

                    val notificationDateTime = createNotificationDateTime(
                        precedingPeriod,
                        it.key,
                        today,
                        notiTime
                    )

                    alarmManager.set(
                        AlarmManager.RTC_WAKEUP,
                        notificationDateTime.toInstant(zoneOffset).toEpochMilli(),
                        alarmIntent
                    )
                    Log.d(TAG, "setAlarms: Alarm set @ $notificationDateTime")
                }
            }
        }
        return alarmId
    }

    private fun createBroadcastNotificationIntent(
        ctx: Context,
        isStart: Boolean,
        alarmId: Int,
        @StringRes titleRes: Int,
        notificationText: String
    ): PendingIntent? {
        val intent = Intent(ctx, Receiver::class.java).apply {
            putExtra("type", if (isStart) "start" else "end")
            putExtra("reqCode", alarmId)
            putExtra("titleRes", titleRes)
            putExtra("text", notificationText)
        }
        return PendingIntent.getBroadcast(
            ctx,
            alarmId,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT
        )
    }

    private fun createNotificationDateTime(
        preNotiTimePeriod: PreNotiTimePeriod,
        monthDay: MonthDay,
        today: LocalDate,
        notiTime: LocalTime
    ): LocalDateTime {
        return when (preNotiTimePeriod) {
            PreNotiTimePeriod.AtDay ->
                monthDay.atYear(today.year)
            PreNotiTimePeriod.WeekBefore ->
                monthDay.atYear(today.year).minusWeeks(1)
            PreNotiTimePeriod.MonthBefore ->
                monthDay.atYear(today.year).minusMonths(1)
        }.let { obtainedDate ->
            if (obtainedDate.isBefore(today)) {
                obtainedDate.plusYears(1)
            } else {
                obtainedDate
            }
        }.atTime(notiTime)
    }

    internal fun setAlarms(
        ctx: Context,
        currentDayProvider: TimeDataProvider,
        persistentStorage: PersistentStorage,
        alarmManager: AlarmManager,
        database: Database
    ) {
        val lastCreatedAlarmId = createAlarms(
            ctx,
            persistentStorage,
            alarmManager,
            database,
            currentDayProvider
        )

        clearPreviousAlarms(
            ctx,
            persistentStorage,
            alarmManager,
            lastCreatedAlarmId
        )

        persistentStorage.lastSetAlarmId = lastCreatedAlarmId
        persistentStorage.lastAlarmSetTime = currentDayProvider.getCurrentDay().toEpochDay()
    }

    private fun createAlarms(
        ctx: Context,
        persistentStorage: PersistentStorage,
        alarmManager: AlarmManager,
        database: Database,
        currentDayProvider: TimeDataProvider
    ): Int {
        val allItems = database.allFruits + database.allVegetables
        val startMap = groupByStartDates(allItems)
        val endMap = groupByEndDates(allItems)
        var reqCode = createAlarmsFor(
            currentDayProvider,
            startMap,
            StartPrecedingDaysObtainer(persistentStorage),
            StartPrecedingDaysMapper(),
            true,
            1,
            ctx,
            persistentStorage,
            alarmManager
        )

        reqCode = createAlarmsFor(
            currentDayProvider,
            endMap,
            EndPrecedingDaysObtainer(persistentStorage),
            EndPrecedingDaysMapper(),
            false,
            reqCode,
            ctx,
            persistentStorage,
            alarmManager
        )
        return reqCode - 1
    }

    private fun clearPreviousAlarms(
        ctx: Context,
        persistentStorage: PersistentStorage,
        alarmManager: AlarmManager,
        lastCreatedAlarmId: Int
    ) {
        val prevMaxReqCode = persistentStorage.lastSetAlarmId
        Log.d(TAG, "setAlarms: Ending with $lastCreatedAlarmId, previous ending $prevMaxReqCode")

        val intent = Intent(ctx, Receiver::class.java)
        for (id in (lastCreatedAlarmId + 1)..prevMaxReqCode) {
            Log.d(TAG, "setAlarms: Cancelling an alarm of id = $id")
            val alarmIntent = PendingIntent.getBroadcast(
                ctx,
                id,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT
            )
            alarmManager.cancel(alarmIntent)
        }
    }

    private fun groupByStartDates(items: List<FoodItem>): Map<MonthDay, List<FoodItem>> {
        val itemsByStartDay1 = items.filter { it.startDay1 != null }.groupBy { it.startDay1!! }
        val itemsByStartDay2 = items.filter { it.startDay2 != null }.groupBy { it.startDay2!! }

        return itemsByStartDay1 + itemsByStartDay2
    }

    private fun groupByEndDates(items: List<FoodItem>): Map<MonthDay, List<FoodItem>> {
        val itemsByEndDay1 = items.filter { it.endDay1 != null }.groupBy { it.endDay1!! }
        val itemsByEndDay2 = items.filter { it.endDay2 != null }.groupBy { it.endDay2!! }

        return itemsByEndDay1 + itemsByEndDay2
    }
}
